File: Tuning/sample_code/05_mattp/src/fortran/main.f90
! Copyright 2024 Research Organization for Information Science and Technology
!-----------------------------------------------------------------------
!  Loop blocking: Matrix transpose 
!  Author:      Yukihiro Ota (yota@rist.or.jp)
!  Reference:   See Sec.3.4 in Introduction to High Performance 
!               Computing for Scientists and Engineers by 
!               Hager and Wellein (CRC Press, 2011).
!-----------------------------------------------------------------------
program main
  implicit none

  integer :: nrow, ncol, nbk, nsize, nrbk, ncbk
  integer :: ierr, hnsize, idm 
  integer :: i, j, itr

  character(len=32) :: cbuff

  real(8) :: wk, mema
  real(8),allocatable :: a(:,:), b(:,:)

#ifdef __USE_CPU_TIMER__
  real(8) :: elpsta, elp(2) 
#endif

  external mattp, mattp_bk 

  !---------------------------------------------------------------------
  !  get matrix row size and block size from arguments 
  !---------------------------------------------------------------------
  ierr = command_argument_count ()
  if ( ierr /= 2 ) then 
    write (6,'(1a48)') "[usage] ./run.x $1 $2                          " 
    write (6,'(1a48)') "       $1: matrix row size (integer)           "
    write (6,'(1a48)') "       $2: block size      (integer)           "
    stop
  endif
  !! * argument 1
  call get_command_argument ( number=1, value=cbuff )
  read (cbuff,'(1i10)') nrow
  ncol = nrow
  if ( nrow < 0 ) then
    write (6,'(1a48)') "Matrix row size is greater than 0              "
    write (6,'(1a48)') "The program is terminated.                     "
    stop
  endif
  if ( nrow > 8192 ) then
    !! 2 double arrays * 8 bytes * (nrow)^2 < 1 GiB=2^30 bytes */
    write (6,'(1a48)') "Matrix row size must be smaller than 8192.     "
    write (6,'(1a48)') "The program is terminated.                     "
    stop
  endif
  !! * argument 2
  call get_command_argument ( number=2, value=cbuff )
  read (cbuff,'(1i10)') nbk 
  if ( nbk <= 0 ) then
    write (6,'(1a48)') "Block size must be strictly positive.          "
    write (6,'(1a48)') "The program is terminated.                     "
    stop
  endif
  if ( nbk >= nrow ) then
    write (6,'(1a48)') "Block size must be smaller than matrix row size"
    write (6,'(1a48)') "The program is terminated.                     "
    stop
  endif
  nrbk = (nrow + (nbk-1) ) /nbk !! block num. (rounded up)
  ncbk = (ncol + (nbk-1) ) /nbk !! block num. (rounded up)
  !---------------------------------------------------------------------
  !  memory allocation (and obtain physical memory) 
  !---------------------------------------------------------------------
  nsize = nrow * ncol
  allocate ( a(nrow,ncol) )
  allocate ( b(nrow,ncol) )
  wk = 1.d0 / nsize
  hnsize = nsize / 2
  do j = 1, ncol
    do i = 1, nrow
      a(i,j) = 0.d0
      b(i,j) = 1.d0 + (i+j*nrow-hnsize)*wk 
    enddo
  enddo
  !---------------------------------------------------------------------
  !  matrix transpose
  !---------------------------------------------------------------------
  mema = 2.d0*8.d0*nsize*1.0d-6 !! MB
  idm = 0
#ifdef __USE_CPU_TIMER__
  call cpu_time ( elpsta ) 
#endif 
  do itr = 1, 10000
    call mattp ( b, a, nrow, ncol, itr, idm )
    idm = idm + 1
  enddo
#ifdef __USE_CPU_TIMER__
  call cpu_time ( elp(1) )
  elp(1) = elp(1) - elpsta
#endif
  !---------------------------------------------------------------------
  !  standard output 
  !---------------------------------------------------------------------
  write (6,'(1a48)') "Matrix transpose (standard)                      "
  write (6,'(1a48)') "Number of iterations                             "
  write (6,'(1i10)') idm
  write (6,'(1a48)') "Check matrix transpose                           "
  write (6,'(2i5,2f17.6)') 1, ncol, a(1,ncol), b(1,ncol)
  write (6,'(2i5,2f17.6)') nrow, 1, a(nrow,1), b(nrow,1)
  write (6,'(1a48)') "Req. memory (MB)                                 "
  write (6,'(1f17.6)') mema
#ifdef __USE_CPU_TIMER__
  write (6,'(1a48)') "CPU time (sec)                                   "
  write (6,'(1f17.6)') elp(1) 
#endif
  !---------------------------------------------------------------------
  !  memory deallocation and allocation (and obtain physical memory) 
  !---------------------------------------------------------------------
  deallocate ( a )
  deallocate ( b )
  allocate ( a(nrow,ncol) )
  allocate ( b(nrow,ncol) )
  wk = 1.d0 / nsize
  hnsize = nsize / 2
  do j = 1, ncol
    do i = 1, nrow
      a(i,j) = 0.d0
      b(i,j) = 1.d0 + (i+j*nrow-hnsize)*wk 
    enddo
  enddo
  !---------------------------------------------------------------------
  !  matrix transpose  (blocking)
  !---------------------------------------------------------------------
  mema = 2.d0*8.d0*nsize*1.0d-6 !! MB
  idm = 0
#ifdef __USE_CPU_TIMER__
  call cpu_time ( elpsta ) 
#endif 
  do itr = 1, 10000
    call mattp_bk ( b, a, nrow, ncol, itr, idm, nbk )
    idm = idm + 1
  enddo
#ifdef __USE_CPU_TIMER__
  call cpu_time ( elp(2) )
  elp(2) = elp(2) - elpsta
#endif
  !---------------------------------------------------------------------
  !  standard output 
  !---------------------------------------------------------------------
  write (6,'(1a48)') "Matrix transpose (blocking)                      "
  write (6,'(1a48)') "Number of iterations                             "
  write (6,'(1i10)') idm
  write (6,'(1a48)') "Check matrix transpose                           "
  write (6,'(2i5,2f17.6)') 1, ncol, a(1,ncol), b(1,ncol)
  write (6,'(2i5,2f17.6)') nrow, 1, a(nrow,1), b(nrow,1)
  write (6,'(1a48)') "Req. memory (MB)                                 "
  write (6,'(1f17.6)') mema
  write (6,'(1a48)') "Number of blocks w.r.t. row and column           "
  write (6,'(4x,1i5,2x,1i5)') nrbk, ncbk
#ifdef __USE_CPU_TIMER__
  write (6,'(1a48)') "CPU time (sec)                                   "
  write (6,'(1f17.6)') elp(2) 
#endif
  !---------------------------------------------------------------------
  !  Finalize 
  !---------------------------------------------------------------------
  deallocate ( a, b ) 
  stop
end 
!-----------------------------------------------------------------------
!  mattp 
!-----------------------------------------------------------------------
subroutine mattp ( a, a_tp, nr, nc, it, idm )
  implicit none

  integer,intent(in) :: nr, nc, it
  real(8),intent(in) :: a(nr,nc)

  integer,intent(out) :: idm
  real(8),intent(out) :: a_tp(nr,nc)
  !---------------------------------------------------------------------
  ! * local variables
  !---------------------------------------------------------------------
  integer :: i, j
  !---------------------------------------------------------------------
  ! * main routine 
  !---------------------------------------------------------------------
  do j = 1, nc
  do i = 1, nr

    a_tp(i,j) = a(j,i) !! 1 load, 1 sotre, no FP operation

  enddo
  enddo

  idm = it - 1

  return
end subroutine
!-----------------------------------------------------------------------
!  blocked mattp 
!-----------------------------------------------------------------------
subroutine mattp_bk ( a, a_tp, nr, nc, it, idm, nbk )
  implicit none

  integer,intent(in) :: nr, nc, it, nbk
  real(8),intent(in) :: a(nr,nc)

  integer,intent(out) :: idm
  real(8),intent(out) :: a_tp(nr,nc)
  !---------------------------------------------------------------------
  ! * local variables
  !---------------------------------------------------------------------
  integer :: i, j, ii, jj
  !---------------------------------------------------------------------
  ! * main routine 
  !---------------------------------------------------------------------
  do jj = 1, nc, nbk
  do ii = 1, nr, nbk

    do j = jj, min(jj+nbk-1,nc)
    do i = ii, min(ii+nbk-1,nr)
      a_tp(i,j) = a(j,i)  !! 1 load, 1 store, no FP operation
    enddo
    enddo

  enddo
  enddo

  idm = it - 1

  return
end subroutine

File: Tuning/sample_code/05_mattp/src/c/main.c
/* Copyright 2024 Research Organization for Information Science and Technology */
/*---------------------------------------------------------------------
 *  Loop blocking: Matrix transpose 
 *  Author:      Yukihiro Ota (yota@rist.or.jp)
 *  Reference:   See Sec.3.4 in Introduction to High Performance 
 *               Computing for Scientists and Engineers by 
 *               Hager and Wellein (CRC Press, 2011).
 *-------------------------------------------------------------------*/
#include <stdio.h>
#include <stdlib.h>
#include "timer.h"

#define MIN(a, b) ( (a) < (b) ? (a) : (b) )

void mattp ( double *a, double *a_tp, const int nr, const int nc, 
             int it, int *idm ) ; 
void mattp_bk ( double *a, double *a_tp, const int nr, const int nc, 
                int it, int *idm, const int nbk ) ; 

int main ( int argc, char* argv[] ) 
{
  int nrow, ncol, nbk, nsize, nrbk, ncbk;
  int idm ;

  double mema ;
  double *a ;
  double *b ;

#ifdef __USE_CPU_TIMER__
  double elpsta, elp[2] ;
#endif

  /*-------------------------------------------------------------------
   * get matrix row size and block size from arguments 
   *-----------------------------------------------------------------*/
  if ( argc != 3 ) {
    printf ("[usage] run.x $1 $2\n") ;
    printf ("        $1: matrix row size (integer)\n") ;
    printf ("        $2: block size      (integer)\n") ;
    return EXIT_SUCCESS ;
  }
  // argument 1
  nrow = atoi (argv[1]) ;
  ncol = nrow ;
  if ( nrow <= 0 ) {
    printf ("Matrix dimension must be strictly positive\n") ;
    printf ("The program is terminated\n") ;
    return EXIT_FAILURE ;
  }
  if ( nrow > 8192 ) {
    /* 2 double arrays * 8 bytes * (nrow)^2 < 1 GiB=2^30 bytes */
    printf ("Matrix row size must be smaller than 8192\n") ;
    printf ("The program is terminated\n") ;
    return EXIT_FAILURE ;
  } 
  // argument 2
  nbk = atoi (argv[2]) ;
  if ( nbk <= 0 ) {
    printf ("Block size must be strictly positive\n") ;
    printf ("The program is terminated\n") ;
    return EXIT_FAILURE ;
  }
  if ( nbk >= nrow ) {
    printf ("Block size must be smaller than matrix row size\n") ;
    printf ("The program is terminated\n") ;
    return EXIT_FAILURE ;
  }
  nrbk = (nrow + (nbk - 1) ) /nbk ; // Block num. (rounded up)
  ncbk = (ncol + (nbk - 1) ) /nbk ; // Block num. (rounded up)
  /*-------------------------------------------------------------------
   * memory allocation  (and obtain physical memory)
   *-----------------------------------------------------------------*/
  nsize = nrow * ncol ;
  a = (double *)malloc ( sizeof(double)*nsize ) ;
  b = (double *)malloc ( sizeof(double)*nsize ) ;
  {
    double wk = 1.0/nsize ;
    int hnsize = nsize / 2 ; //use of bitwise op would be preferable .. 
    for ( int i=0; i<nrow; ++i ) {
      for ( int j=0; j<ncol; ++j ) {
        int n = j + i*ncol ;
	a[n] = 0.0 ;
        b[n] = 1.0 + (n-hnsize)*wk ; 
      }
    }
  }
  /*-------------------------------------------------------------------
   * matrix transpose 
   *-----------------------------------------------------------------*/
  mema = 2.0*sizeof(double)*nsize*1.0e-6 ; // MB
  idm = 0 ;
#ifdef __USE_CPU_TIMER__
  elpsta = get_cpu_time () ; 
#endif
  for ( int itr=0; itr<10000; ++itr) {
    mattp ( b, a, nrow, ncol, itr, &idm ) ; // a = ^t b
    idm += 1 ;
  }
#ifdef __USE_CPU_TIMER__
  elp[0] = get_cpu_time () ; 
  elp[0] = elp[0] - elpsta ;
#endif
  /*-------------------------------------------------------------------
   * standard output 
   *-----------------------------------------------------------------*/
  printf ("Matrix transpose (standard)    \n") ;
  printf ("Number of iterations           \n") ;
  printf ("%d\n", idm ) ;
  printf ("Check matrix transpose         \n") ;
  printf ("%d %d %17.6f %17.6f\n", 
          0, ncol-1, a[ncol-1], b[ncol-1]);
  printf ("%d %d %17.6f %17.6f\n", 
          nrow-1, 0, a[(nrow-1)*ncol], b[(nrow-1)*ncol]);
  printf ("Req. memory (MB)               \n") ;
  printf ("%17.6f\n", mema) ;
#ifdef __USE_CPU_TIMER__
  printf ("CPU time (sec)                 \n") ;
  printf ("%17.6f\n", elp[0]) ;
#endif
  /*-------------------------------------------------------------------
   * memory delete and allocation  (and obtain physical memory)
   *-----------------------------------------------------------------*/
  free ( a ); a = NULL;
  free ( b ); b = NULL;
  a = (double *)malloc ( sizeof(double)*nsize ) ;
  b = (double *)malloc ( sizeof(double)*nsize ) ;
  {
    double wk = 1.0/nsize ;
    int hnsize = nsize / 2 ; //use of bitwise op would be preferable .. 
    for ( int i=0; i<nrow; ++i ) {
      for ( int j=0; j<ncol; ++j ) {
        int n = j + i*ncol ;
	a[n] = 0.0 ;
        b[n] = 1.0 + (n-hnsize)*wk ; 
      }
    }
  }
  /*-------------------------------------------------------------------
   * matrix transpose (blocking)
   *-----------------------------------------------------------------*/
  mema = 2.0*sizeof(double)*nsize*1.0e-6 ; // MB
  idm = 0 ;
#ifdef __USE_CPU_TIMER__
  elpsta = get_cpu_time () ; 
#endif
  for ( int itr=0; itr<10000; ++itr) {
    mattp_bk ( b, a, nrow, ncol, itr, &idm, nbk ) ; // a = ^t b
    idm += 1 ;
  }
#ifdef __USE_CPU_TIMER__
  elp[1] = get_cpu_time () ; 
  elp[1] = elp[1] - elpsta ;
#endif
  /*-------------------------------------------------------------------
   * standard output 
   *-----------------------------------------------------------------*/
  printf ("Matrix transpose (blocking)    \n") ;
  printf ("Number of iterations           \n") ;
  printf ("%d\n", idm ) ;
  printf ("Check matrix transpose         \n") ;
  printf ("%d %d %17.6f %17.6f\n", 
          0, ncol-1, a[ncol-1], b[ncol-1]);
  printf ("%d %d %17.6f %17.6f\n", 
          nrow-1, 0, a[(nrow-1)*ncol], b[(nrow-1)*ncol]);
  printf ("Req. memory (MB)               \n") ;
  printf ("%17.6f\n", mema) ;
  printf ("Number of blocks w.r.t.        \n") ;
  printf ("   row: %d  column: %d\n", nrbk, ncbk ) ;
#ifdef __USE_CPU_TIMER__
  printf ("CPU time (sec)                 \n") ;
  printf ("%17.6f\n", elp[1]) ;
#endif
  /*-------------------------------------------------------------------
   * finalization  
   *-----------------------------------------------------------------*/
  free ( a ) ; free ( b ) ;

  return EXIT_SUCCESS ;
}
/*---------------------------------------------------------------------
 * mattp
 *-------------------------------------------------------------------*/
void mattp ( double *a, double *a_tp, const int nr, const int nc, 
             int it, int *idm ) {

  for ( int i=0; i<nr; ++i ) {
  for ( int j=0; j<nc; ++j ) {
      int id  = j + i*nc ; 
      int idt = i + j*nr ;  // stride nr access on loop j
      a_tp[id] = a[idt] ;   // 1 load, 1 store, no FP operation
  }}

  *idm = it ;
}     
/*---------------------------------------------------------------------
 * blocked mattp
 *-------------------------------------------------------------------*/
void mattp_bk ( double *a, double *a_tp, const int nr, const int nc, 
                int it, int *idm, const int nbk ) {

  for ( int ii=0; ii<nr; ii+=nbk ) { 
    int iend = MIN(ii+nbk,nr) ; 
  for ( int jj=0; jj<nc; jj+=nbk ) {
    int jend = MIN(jj+nbk,nc) ; 

    for ( int i=ii; i<iend; ++i ) { 
    for ( int j=jj; j<jend; ++j ) {
       int id  = j + i*nc ; // What about stride? 
       int idt = i + j*nr ; // What about stride?
       a_tp[id] = a[idt] ;     // 1 load, 1 store, no FP operation
    }}

  }}

  *idm = it ;
}     

